File: Data Structures & Algorithms/valid-tree/submission-7.js
class Solution {
  /**
   * @param {number} n
   * @param {number[][]} edges
   * @returns {boolean}
   */
  validTree(n, edges) {
    if (!edges.length) return true;
    const adj = {};
    const vis = new Set();
    for (const [n1, n2] of edges) {
      adj[n1] ? adj[n1].push(n2) : (adj[n1] = [n2]);
      adj[n2] ? adj[n2].push(n1) : (adj[n2] = [n1]);
    }

    return dfs(0, -1) && vis.size == n;

    function dfs(i, prev) {
      if (vis.has(i)) return false;
      vis.add(i);
      for (const j of adj[i]) {
        if (prev === j) continue;
        if (!dfs(j, i)) return false;
      }
      return true;
    }
  }
}

File: Data Structures & Algorithms/count-connected-components/submission-7.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        l={i:[] for i in range(n)}
        for i,j in edges:
            l[i].append(j)
            l[j].append(i)
        res=[0]
        v=[]
        def f(n,ic=False):
            if n in v:return
            if ic:res[0]+=1
            v.append(n)
            for nn in l[n]:f(nn,)
        for n in l:
            f(n,ic=True)
        return res[0]